### ISSUE TITLE
Assets are not found with Play on IDEA when running integration tests

### ISSUE THREAD

**Issue by jarmo** [playframework/playframework#3234]
- Create new project with Play 2.3.2 `activator` for IDEA:

```
> activator new my-first-app play-java
> cd my-first-app
> activator idea
```
- Open `my-first-app` with IDEA and run all tests.
- Tests pass, but this is written to `application.log`:

```
2014-07-30 17:25:05,231 - [ERROR] - from com.gargoylesoftware.htmlunit.html.HtmlPage in main 
Error loading JavaScript from [http://localhost:3333/assets/javascripts/hello.js].
java.io.IOException: Unable to download JavaScript from 'http://localhost:3333/assets/javascripts/hello.js' (status 404).
    at com.gargoylesoftware.htmlunit.html.HtmlPage.loadJavaScriptFromUrl(HtmlPage.java:1106) ~[htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HtmlPage.loadExternalJavaScriptFile(HtmlPage.java:1039) ~[htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HtmlScript.executeScriptIfNeeded(HtmlScript.java:409) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HtmlScript$3.execute(HtmlScript.java:266) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HtmlScript.onAllChildrenAddedToPage(HtmlScript.java:286) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:683) [htmlunit-2.13.jar:2.13]
    at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source) [xercesImpl-2.11.0.jar:na]
    at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:642) [htmlunit-2.13.jar:2.13]
    at org.cyberneko.html.HTMLTagBalancer.callEndElement(HTMLTagBalancer.java:1170) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.HTMLTagBalancer.endElement(HTMLTagBalancer.java:1072) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.filters.DefaultFilter.endElement(DefaultFilter.java:206) [nekohtml-1.9.19.jar:na]
    at org.cyberneko.html.filters.NamespaceBinder.endElement(NamespaceBinder.java:330) [nekohtml-1.9.19.jar:na]
    at org.cyberneko.html.HTMLScanner$ContentScanner.scanEndElement(HTMLScanner.java:3116) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.HTMLScanner$ContentScanner.scan(HTMLScanner.java:2083) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.HTMLScanner.scanDocument(HTMLScanner.java:918) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.HTMLConfiguration.parse(HTMLConfiguration.java:499) [nekohtml-1.9.19.jar:1.9.19]
    at org.cyberneko.html.HTMLConfiguration.parse(HTMLConfiguration.java:452) [nekohtml-1.9.19.jar:1.9.19]
    at org.apache.xerces.parsers.XMLParser.parse(Unknown Source) [xercesImpl-2.11.0.jar:na]
    at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.parse(HTMLParser.java:899) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HTMLParser.parse(HTMLParser.java:242) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.html.HTMLParser.parseHtml(HTMLParser.java:188) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.DefaultPageCreator.createHtmlPage(DefaultPageCreator.java:268) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.DefaultPageCreator.createPage(DefaultPageCreator.java:156) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseInto(WebClient.java:437) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:311) [htmlunit-2.13.jar:2.13]
    at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:376) [htmlunit-2.13.jar:2.13]
    at org.openqa.selenium.htmlunit.HtmlUnitDriver.get(HtmlUnitDriver.java:474) [selenium-htmlunit-driver-2.39.0.jar:na]
    at org.openqa.selenium.htmlunit.HtmlUnitDriver.get(HtmlUnitDriver.java:463) [selenium-htmlunit-driver-2.39.0.jar:na]
    at org.fluentlenium.core.Fluent.goTo(Fluent.java:336) [fluentlenium-core-0.9.2.jar:na]
    at IntegrationTest$1.invoke(IntegrationTest.java:22) [test-classes/:na]
    at IntegrationTest$1.invoke(IntegrationTest.java:20) [test-classes/:na]
    at play.test.Helpers.running(Helpers.java:549) [play-test_2.11-2.3.2.jar:2.3.2]
    at play.test.Helpers.running(Helpers.java:536) [play-test_2.11-2.3.2.jar:2.3.2]
    at IntegrationTest.test(IntegrationTest.java:20) [test-classes/:na]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_51]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_51]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_51]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_51]
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47) [junit-4.11.jar:na]
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.11.jar:na]
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44) [junit-4.11.jar:na]
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271) [junit-4.11.jar:na]
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70) [junit-4.11.jar:na]
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309) [junit-4.11.jar:na]
    at org.junit.runners.Suite.runChild(Suite.java:127) [junit-4.11.jar:na]
    at org.junit.runners.Suite.runChild(Suite.java:26) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229) [junit-4.11.jar:na]
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309) [junit-4.11.jar:na]
    at org.junit.runner.JUnitCore.run(JUnitCore.java:160) [junit-4.11.jar:na]
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:74) [junit-rt.jar:na]
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:211) [junit-rt.jar:na]
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:67) [junit-rt.jar:na]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_51]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_51]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_51]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_51]
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134) [idea_rt.jar:na]
```

If writing more complex integration tests, where JavaScript is actually needed, then running tests is impossible with `IDEA`. Play 2.2 works well. Also, when running `activator test` then everything works as expected.

**Comment by albuch** [comment]
So @jroper what would be the suggested pre launch SBT actions that you would suggest for an integration test in IDEA? 
I'm using [idea-sbt-plugin](https://github.com/orfjackal/idea-sbt-plugin/) to avoid full Makes on run and test configurations already and can add additional actions.
Right now I'm using `;test-compile;web-test:assets` along with `managedResourceDirectories in Test <+= baseDirectory ( _ /"target/web/public/test" )` which seems to do the trick.
Anything else to add and what ordering should be used?

**Comment by wsargent** [comment]
Is this still an issue?

**Comment by kiruSan** [comment]
@albuch What is the proper configuration for idea-sbt-plugin? Could you provide sample code?

**Comment by wsargent** [comment]
I don't think this is a real issue any more, since we're not using sbt-idea any more, and are instead using "Import Project" directly from IntelliJ IDEA 15 (and IntelliJ is smart enough to understand the older sbt-idea format and remind you to upgrade)

**Comment by wsargent** [comment]
Closing (comments still enabled for those who want to discuss 2.3.x)

**Comment by kiruSan** [comment]
@wsargent  Thank you for commenting on this issue.
I don't think it is fixed, this issue was created for play 2.3.x but I still get the error with the latest Play (2.5.2) and IntellJ version ( 2016.1.1 )

To reproduce:

`activator new my-first-app play-java`
=> In IntellJ: File -> Open -> select my-first-app
=> Run IntegrationTest

I get the following exception: 
```

[info] application - Creating Pool for datasource 'default'
[info] application - ApplicationTimer demo: Starting application at 2016-04-16T09:19:54.702Z
[error] c.g.h.h.HtmlPage - Error loading JavaScript from [http://localhost:3333/assets/javascripts/hello.js].
java.io.IOException: Unable to download JavaScript from 'http://localhost:3333/assets/javascripts/hello.js' (status 404).
	at com.gargoylesoftware.htmlunit.html.HtmlPage.loadJavaScriptFromUrl(HtmlPage.java:1055)
	at com.gargoylesoftware.htmlunit.html.HtmlPage.loadExternalJavaScriptFile(HtmlPage.java:982)
	at com.gargoylesoftware.htmlunit.html.HtmlScript.executeScriptIfNeeded(HtmlScript.java:352)
	at com.gargoylesoftware.htmlunit.html.HtmlScript$2.execute(HtmlScript.java:238)
	at com.gargoylesoftware.htmlunit.html.HtmlScript.onAllChildrenAddedToPage(HtmlScript.java:257)
	at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:747)
	at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source)
	at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:704)
	at org.cyberneko.html.HTMLTagBalancer.callEndElement(HTMLTagBalancer.java:1170)
	at org.cyberneko.html.HTMLTagBalancer.endElement(HTMLTagBalancer.java:1072)
[info] application - ApplicationTimer demo: Stopping application at 2016-04-16T09:19:57.496Z after 3s.
[info] application - Shutting down connection pool.
```

**Comment by llb4ll** [comment]
Same issue here: IntelliJ IDEA 2016.1.13 + Play 2.5.6

**Comment by ashleymercer** [comment]
I've raised this as a bug in IntelliJ since the problem is really there: the IntelliJ test runners are not properly obeying `fullClasspath in Test` - even their example test classes in a brand new Play project don't work out of the box!

**Comment by wsargent** [comment]
Okay, closing as this is an IDEA bug.

**Comment by rmmeans** [comment]
For others that are still finding this thread and having problems with assets being available in test runs we found success on Play 2.6 for our functional tests ( our functional tests are just another suite that extends the `Test` configuration: ```lazy val FunctionalTest = config("functional") extend Test``` by adding this one line:
```scala
(managedClasspath in FunctionalTest) += (packageBin in Assets).value
```

This was hidden in the details of the `sbt-web` README. We were previously trying to use what was suggested in here with `web-test:assets` and `managedResourceDirectories in FunctionalTest <+= baseDirectory ( _ /"target/web/public/test" )` but still not having luck when our CI job would run. 

That one line was all we needed and now everything is working as desired in our test runs!
